File: backend/app/models/screening.py
from __future__ import annotations

from enum import StrEnum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Finding(StrEnum):
    NO_ISSUE = "no_issue_identified_within_scoped_rules"
    WITHIN_LIMIT = "restriction_within_limit"
    RESTRICTION_EXCEEDED = "restriction_exceeded"
    PROHIBITED = "prohibited_substance_identified"
    INFORMATION_MISSING = "information_missing"
    IDENTITY_UNRESOLVED = "identity_unresolved"
    PROFESSIONAL_REVIEW = "professional_review_required"


class IdentityStatus(StrEnum):
    RESOLVED = "resolved"
    UNRESOLVED = "unresolved"
    AMBIGUOUS = "ambiguous"
    REVIEW_REQUIRED = "review_required"


class ConcentrationInput(StrictModel):
    value: float = Field(ge=0, allow_inf_nan=False)
    unit: str = Field(min_length=1)
    basis: str | None
    preparation_stage: str = Field(min_length=1)

    @field_validator("unit", "preparation_stage")
    @classmethod
    def require_nonblank_semantics(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("Concentration unit and preparation stage cannot be blank")
        return value

    @field_validator("basis")
    @classmethod
    def reject_blank_basis(cls, value: str | None) -> str | None:
        if value is not None and not value.strip():
            raise ValueError("Concentration basis must be null or non-blank")
        return value


class IngredientInput(StrictModel):
    name: str | None = None
    cas_number: str | None = None
    concentration: ConcentrationInput | None = None

    @model_validator(mode="after")
    def require_identity_input(self) -> "IngredientInput":
        if not (self.name and self.name.strip()) and not (self.cas_number and self.cas_number.strip()):
            raise ValueError("At least one of ingredient name or CAS number is required")
        return self


class IdentityCandidate(StrictModel):
    substance_id: str
    source_document: str
    regulatory_section: str
    reference_number: str
    original_substance_name: str
    normalized_substance_name: str
    cas_numbers: list[str] = Field(default_factory=list)
    match_methods: list[str] = Field(default_factory=list)
    cross_reference_statuses: list[str] = Field(default_factory=list)


class IdentityResolution(StrictModel):
    status: IdentityStatus
    match_methods: list[str] = Field(default_factory=list)
    singapore_candidates: list[IdentityCandidate] = Field(default_factory=list)
    acd_candidates: list[IdentityCandidate] = Field(default_factory=list)
    resolved_singapore_substance_id: str | None = None
    reasons: list[str] = Field(default_factory=list)


class RuleEvidence(StrictModel):
    dataset_version: str
    rule_id: str
    source_document: str
    source_version: str | None
    effective_date: str | None
    document_revision: str | None
    retrieval_date: str
    source_url: str
    regulatory_section: str
    reference_number: str
    substance_name: str
    product_context: str | None
    concentration: dict[str, Any] | None
    other_conditions: str | None
    required_warning: str | None
    source_text: str
    source_pages: list[int]
    normalization_status: str
    review_reasons: list[str]
    raw_record_id: str
    raw_fragments: list[dict[str, Any]] = Field(default_factory=list)
    regulation_6_provisions: list[dict[str, Any]] = Field(default_factory=list)
    cross_references: list[dict[str, Any]] = Field(default_factory=list)
    acd_counterpart_rules: list[dict[str, Any]] = Field(default_factory=list)


class RuleEvaluation(StrictModel):
    rule_id: str
    evaluation_status: str
    finding: Finding | None = None
    reasons: list[str] = Field(default_factory=list)
    submitted_concentration: ConcentrationInput | None = None
    evidence: RuleEvidence


class ScreeningResult(StrictModel):
    dataset_version: str
    accepted_baseline_sha256: str
    submitted_ingredient: IngredientInput
    submitted_product_context: str | None
    identity: IdentityResolution
    primary_finding: Finding
    confirmed_findings: list[Finding] = Field(default_factory=list)
    review_required: bool
    review_reasons: list[str] = Field(default_factory=list)
    rule_evaluations: list[RuleEvaluation] = Field(default_factory=list)
    inactive_evidence: list[RuleEvidence] = Field(default_factory=list)
    searched_singapore_parts: list[str] = Field(default_factory=list)
    scope_note: str = (
        "This is an initial Singapore screening result within the implemented rules only; "
        "professional review remains required."
    )

File: frontend/src/types/screening.ts
export type ConcentrationUnit = "percent" | "mg/kg" | "ppm";
export type ConcentrationBasis =
  | "NH3"
  | "free base"
  | "zinc"
  | "sulphate"
  | "hydrochloride"
  | "tetrahydrochloride";
export type PreparationStage = "finished_product" | "after_mixing" | "ready_for_use";

export type Finding =
  | "no_issue_identified_within_scoped_rules"
  | "restriction_within_limit"
  | "restriction_exceeded"
  | "prohibited_substance_identified"
  | "information_missing"
  | "identity_unresolved"
  | "professional_review_required";

export interface Concentration {
  value: number;
  unit: ConcentrationUnit;
  basis: ConcentrationBasis | null;
  preparation_stage: PreparationStage;
}

export interface FormulationIngredient {
  name: string;
  cas_number: string | null;
  concentration: Concentration | null;
}

export interface FormulationRequest {
  formulation_id: string | null;
  formulation_name: string | null;
  product_context: string | null;
  ingredients: FormulationIngredient[];
}

export interface ScreeningOptions {
  jurisdiction: "Singapore";
  dataset_version: string;
  accepted_baseline_sha256: string;
  product_contexts: string[];
  concentration_units: ConcentrationUnit[];
  concentration_bases: Array<ConcentrationBasis | null>;
  preparation_stages: PreparationStage[];
}

export interface SourceSnapshot {
  source_document: string;
  title: string;
  authority: string;
  sha256: string;
  page_count: number;
  source_url: string;
  document_revision: string | null;
  effective_date: string | null;
  retrieval_date: string;
  snapshot_generated_at: string | null;
}

export interface IdentityCandidate {
  substance_id: string;
  source_document: string;
  regulatory_section: string;
  reference_number: string;
  original_substance_name: string;
  normalized_substance_name: string;
  cas_numbers: string[];
  match_methods: string[];
  cross_reference_statuses: string[];
}

export interface IdentityResolution {
  status: "resolved" | "unresolved" | "ambiguous" | "review_required";
  match_methods: string[];
  singapore_candidates: IdentityCandidate[];
  acd_candidates: IdentityCandidate[];
  resolved_singapore_substance_id: string | null;
  resolved_singapore_substance_ids: string[];
  identity_source_type: "singapore_regulatory_source" | "acd_regulatory_source" | "ingredient_identity_catalogue" | null;
  identity_source_name: string | null;
  singapore_linkage_status: "not_applicable" | "linked" | "verified_not_represented" | "unresolved";
  catalogue_identity: CatalogueIdentity | null;
  linkage_evidence: LinkageEvidence | null;
  reasons: string[];
}

export interface LinkageTarget {
  raw_record_id: string;
  rule_id: string;
  substance_id: string;
  part: string;
  reference: string;
  source_substance_name: string;
  source_document: string;
  source_hash: string;
}

export interface LinkageEvidence {
  linkage_id: string;
  accepted_status: "linked" | "verified_not_represented" | "unresolved";
  applicable_to_active_baseline: boolean;
  identity_dataset_version: string;
  identity_dataset_hash: string;
  singapore_regulatory_baseline: string;
  singapore_regulatory_baseline_hash: string;
  screened_scope: string[];
  singapore_targets: LinkageTarget[];
  review: {
    reviewed: boolean;
    reviewed_at: string;
    reviewer: string;
    review_basis: string;
    notes: string;
  };
  inapplicability_reasons: string[];
}

export interface CatalogueIdentity {
  ingredient_id: string;
  canonical_name: string;
  source_name: string;
  source_document: string;
  source_version: string;
  source_url: string;
  source_entries: number[];
  source_pages: number[];
  raw_record_ids: string[];
  identity_dataset_version: string;
  accepted_baseline_sha256: string;
}

export interface RawFragment {
  raw_record_id: string;
  source_page: number;
  table_bbox?: number[] | null;
  row_bbox?: number[] | null;
  cells?: Array<string | null>;
  original_row_text?: string;
  is_continuation?: boolean;
  [key: string]: unknown;
}

export interface RegulationProvision {
  provision_id: string;
  regulation: string;
  paragraph: string;
  source_document: string;
  source_pages: number[];
  source_text: string;
  source_version: string | null;
  effective_date: string | null;
  retrieval_date: string;
  source_url: string;
  source_hash: string;
}

export interface CrossReference {
  cross_reference_id: string;
  regulatory_mapping: string;
  reference_number: string;
  acd_rule_ids: string[];
  singapore_rule_ids: string[];
  match_basis: string;
  comparison_status: "aligned" | "changed" | "acd_only" | "singapore_only" | "ambiguous";
  field_differences: Record<string, unknown>;
  candidate_matches: unknown[];
  professional_review_required: boolean;
  review_reasons: string[];
}

export interface CounterpartRule {
  rule_id: string;
  source_document: string;
  source_version: string | null;
  regulatory_section: string;
  reference_number: string;
  substance_name: string;
  cas_numbers: string[];
  source_text: string;
  source_pages: number[];
  source_url: string;
  document_revision: string | null;
  active: boolean;
  normalization_status: string;
  review_reasons: string[];
  raw_fragments: RawFragment[];
  [key: string]: unknown;
}

export interface RuleConcentration {
  value?: number;
  unit?: string;
  comparator?: string;
  basis?: string | null;
  preparation_stage?: string;
  source_field?: string;
  source_text?: string;
  [key: string]: unknown;
}

export interface RuleEvidence {
  dataset_version: string;
  rule_id: string;
  source_document: string;
  source_version: string | null;
  effective_date: string | null;
  document_revision: string | null;
  retrieval_date: string;
  source_url: string;
  regulatory_section: string;
  reference_number: string;
  substance_name: string;
  product_context: string | null;
  concentration: RuleConcentration | null;
  other_conditions: string | null;
  required_warning: string | null;
  source_text: string;
  source_pages: number[];
  normalization_status: string;
  review_reasons: string[];
  raw_record_id: string;
  raw_fragments: RawFragment[];
  regulation_6_provisions: RegulationProvision[];
  cross_references: CrossReference[];
  acd_counterpart_rules: CounterpartRule[];
}

export interface RuleEvaluation {
  rule_id: string;
  evaluation_status: string;
  finding: Finding | null;
  reasons: string[];
  submitted_concentration: Concentration | null;
  evidence: RuleEvidence;
}

export interface IngredientResult {
  dataset_version: string;
  accepted_baseline_sha256: string;
  submitted_ingredient: FormulationIngredient;
  submitted_product_context: string | null;
  identity: IdentityResolution;
  primary_finding: Finding;
  confirmed_findings: Finding[];
  review_required: boolean;
  review_types: Array<"identity_review" | "rule_review">;
  review_reasons: string[];
  rule_evaluations: RuleEvaluation[];
  inactive_evidence: RuleEvidence[];
  searched_singapore_parts: string[];
  scope_note: string;
  submitted_row_number: number;
}

export interface FormulationSummary {
  ingredients_submitted: number;
  prohibited_substance_identified: number;
  restriction_exceeded: number;
  restriction_within_limit: number;
  professional_review_required: number;
  information_missing: number;
  identity_unresolved: number;
  no_issue_identified_within_scoped_rules: number;
  total_requiring_review: number;
  total_unresolved_identities: number;
  duplicate_row_groups: number[][];
}

export interface ScreeningResponse {
  formulation: {
    formulation_id: string | null;
    formulation_name: string | null;
    product_context: string | null;
  };
  dataset: {
    dataset_version: string;
    accepted_baseline_sha256: string;
    sources: SourceSnapshot[];
    identity_catalogue: {
      available: boolean;
      dataset_version: string | null;
      accepted_baseline_sha256: string | null;
      source_name: string | null;
      source_role: string | null;
      ingredient_count: number | null;
      error: string | null;
    };
    identity_linkage: {
      available: boolean;
      dataset_version: string | null;
      accepted_baseline_sha256: string | null;
      identity_dataset_version: string | null;
      singapore_regulatory_baseline: string | null;
      screened_scope: string[];
      accepted_records: number | null;
      linked: number | null;
      verified_not_represented: number | null;
      unresolved: number | null;
      error: string | null;
    };
  };
  summary: FormulationSummary;
  ingredient_results: IngredientResult[];
}

export interface IngredientSearchResult {
  ingredient_id: string;
  canonical_name: string;
  display_name: string;
  identity_source: string;
  source_document: string;
  source_version: string;
  source_entries: number[];
  source_pages: number[];
  raw_record_ids: string[];
}

export interface IngredientSearchResponse {
  query: string;
  dataset_version: string;
  accepted_baseline_sha256: string;
  results: IngredientSearchResult[];
}

export interface EditorIngredient {
  id: string;
  name: string;
  casNumber: string;
  concentrationEnabled: boolean;
  concentrationValue: string;
  unit: ConcentrationUnit;
  basis: ConcentrationBasis | null;
  preparationStage: PreparationStage;
}

export interface FieldIssue {
  path: string;
  message: string;
}
